connector/src/main/scala/com/google/cloud/spark/bigquery/ArrowBinaryIterator.scala

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.spark.bigquery
+
+import java.io.{ByteArrayInputStream, InputStream}
+
+import com.google.cloud.spark.bigquery.converters.ArrowSchemaConverter
+import com.google.protobuf.ByteString
+import org.apache.arrow.vector.VectorSchemaRoot
+import org.apache.arrow.vector.ipc.{ArrowReader, ArrowStreamReader}
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.execution.arrow.ArrowUtils
+import org.apache.spark.sql.vectorized.{ColumnVector, ColumnarBatch}
+
+import collection.JavaConverters._
+
+/**
+ * An iterator for scanning over rows serialized in Arrow format
+ * @param columnsInOrder Sequence of columns in the schema
+ * @param schema Schema in avro format
+ * @param rowsInBytes Rows serialized in binary format for Arrow
+ */
+class ArrowBinaryIterator(columnsInOrder: Seq[String],
+                          schema: ByteString,
+                          rowsInBytes: ByteString) extends Iterator[InternalRow] {
+
+  val allocator = ArrowUtils.rootAllocator.newChildAllocator("ArrowBinaryIterator",
+    0, Long.MaxValue)
+
+  val byteStringWithSchema = schema.concat(rowsInBytes)
+
+  val arrowStreamReader = new ArrowStreamReader(
+    new ByteArrayInputStream(byteStringWithSchema.toByteArray).asInstanceOf[InputStream]
+    , allocator)
+
+  val arrowReaderIterator = new ArrowReaderIterator(arrowStreamReader)
+  val iterator = arrowReaderIterator.flatMap(root => toArrowRows(root, columnsInOrder))
```

Comment:
I don't know if this would change your code design (you might be able to avoid an extra object allocation per row-batch which probably wouldn't change much performance wise) but root in this context will be the same object for every call (each call to loadNextBatch updates the same vectorSchemaRoot on reader).

Reply:
That is fine since root is just an allocator here. But thanks for pointing out! I will keep an eye on it, in case it affects anything.